### DISCUSSION TITLE
Traits are hydrated on mount

### DISCUSSION THREAD

**Discussion by stancl** [livewire/livewire#3130] [7. Bugs (archived)]
### Description

Livewire runs `hydrate{trait name}` on first load, when it should only run `mount{trait name}`.

### Exact steps to reproduce

https://laravelplayground.com/#/snippets/64cccf84-9ee8-431f-a81e-1016e45c4acc

### Stripped-down, copy-pastable code snippets

```php
trait MyTrait
{
    public function hydrateMyTrait()
    {
        dd('hydrating trait');
    }
}

class MyComponent extends Component
{
    use MyTrait;

    public function mount()
    {
        dd('mounting component');
    }
}
```

**Context**
- Livewire version: 2.3.8
- Laravel version: 8.25.0

**Comment by joshhanley** [comment]
@stancl yeah looking at it, it runs `initialise{Trait}` and then `hydrate{Trait}` when 'component.hydrate' is fired (which gets fired on initial and subsequent renders), and then runs `mount{Trait}` when 'component.mount' is fired.

Which is inconsistent with how a components lifecycle methods are fired, which only runs mount on initial render and hydrate on subsequent renders

**Comment by stancl** [comment]
@joshhanley I'm now wondering if this is a bit by design, since Livewire's base `Component` class seems to make big use of traits, so maybe the rationale is that they should be executed before the component itself.

Not sure though.

I also noticed that properties passed

```
@livewire($component, ['here'])
```

are not accessible in trait `mount()`, but are accessible in component `mount()`.

**Comment by stancl** [comment]
@joshhanley I think that `hydrate{Trait}` should run on `component.hydrate.subsequent`? Tbh I'm pretty confused by the hooks, since `component.hydrate.initial` is described as "Run only on the INITIAL hydration (When the component is first loaded)" which to me sounds like `mount`? And the [Reference](https://laravel-livewire.com/docs/2.x/reference) doesn't mention `component.mount`.

It seems like `component.mount` is fired after Component's `mount()` is called, whereas `component.hydrate.initial` is called before that:

```php
return LifecycleManager::fromInitialRequest($name, $id)
    ->initialHydrate() // fires ['component.hydrate', 'component.hydrate.initial']
    ->mount($params) // fires 'component.mount'
    ->renderToView()
    ->initialDehydrate()
    ->toInitialResponse();
```

So it seems that the only way to use `hydrate` as in the component's `hydrate` is to hook into `component.hydrate.initial` in `initialize{Trait}`.

I think the correct behavior is calling `hydrate{Trait}` only when the initial event is fired.

So in `SupportComponentTraits`, this:

```php
$methods = $this->componentIdMethodMap[$component->id]['hydrate'] ?? [];

foreach ($methods as $method) {
    ImplicitlyBoundMethod::call(app(), $method);
}
```

Should be changed to:

```php
Livewire::listen('component.hydrate.initial', function ($component, $params) {
    $methods = $this->componentIdMethodMap[$component->id]['hydrate'] ?? [];

    foreach ($methods as $method) {
        ImplicitlyBoundMethod::call(app(), $method, $params);
    }
});
```

Though there may be breaking changes? Not sure.

The event system seems a bit inconsistent in its naming and maybe that's something that should be changed in v3.

When you look at this:

https://github.com/livewire/livewire/blob/6d0142efaf791b3017335e44e5f433e0bb3169e6/src/HydrationMiddleware/CallHydrationHooks.php#L9-L37

Component's `dehydrate()` is fired both on initial and subsequent dehydrations, whereas `hydrate()` is fired only on the subsequent hydrations. I think that when these two words should just be the opposite of each other, they should have the same logic too.

**Comment by stancl** [comment]
And now I'm puzzled even more.

> component.hydrate | ($component, $request) | Run on EVERY component hydration

This code does not run on the initial load:
```php
Livewire::listen('component.hydrate', fn () => dd('abc'));
```

This code does:
```php
Livewire::listen('component.mount', fn () => dd('abc'));
```

But Livewire should be firing both:

https://github.com/livewire/livewire/blob/6d0142efaf791b3017335e44e5f433e0bb3169e6/src/HydrationMiddleware/CallHydrationHooks.php#L33-L37

So I'm not really sure what's the expected behavior now.

**Comment by joshhanley** [comment]
@stancl ok that's strange I would have thought `component.hydrate` was firing because of that `initialHydrate` method.

The other reason I thought it did fire, if you look at `SupportComponentTraits.php` line 16, it's constructor listens for `component.hydrate`. So maybe if you are trying to listen for that in your trait, the trait won't have been initialised yet, hence why it's not picking it up?
